pkg/util/fxutil/provide_comp.go

```diff
@@ -0,0 +1,381 @@
+// Copyright 2016-present Datadog, Inc.
+
+package fxutil
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	compDef "github.com/DataDog/datadog-agent/comp/def"
+	"go.uber.org/fx"
+)
+
+var (
+	compOutType = reflect.TypeOf((*compDef.Out)(nil)).Elem()
+	compLcType  = reflect.TypeOf((*compDef.Lifecycle)(nil)).Elem()
+	fxInType    = reflect.TypeOf((*fx.In)(nil)).Elem()
+	fxOutType   = reflect.TypeOf((*fx.Out)(nil)).Elem()
+	fxLcType    = reflect.TypeOf((*fx.Lifecycle)(nil)).Elem()
+)
+
+// ProvideComponentConstructor takes as input a Component constructor function
+// that uses plain (non-fx aware) structs as its argument and return value, and
+// returns an fx.Provide'd Option that will properly include that Component
+// into the fx constructor graph.
+//
+// For example, given:
+//
+//	type Provides struct {
+//	    My MyComponent
+//	}
+//	type Requires struct {
+//	    Dep MyDependency
+//	}
+//	func NewComponent(reqs Requires) Provides { ... }
+//
+// then:
+//
+//	ProvideComponentConstructor(NewComponent)
+//
+// will create these anonymous types:
+//
+//	type FxAwareProvides struct {
+//	    fx.Out
+//	    My MyComponent
+//	}
+//	type FxAwareRequires struct {
+//	    fx.In
+//	    Dep MyDependency
+//	}
+//
+// and essentially make this happen:
+//
+//	fx.Provide(func(fxr FxAwareRequires) fxp FxAwareProvides {
+//	    r := makePlainReqs(fxr)
+//	    p := NewComponent(r)
+//	    return makeFxAwareProvides(p)
+//	})
+func ProvideComponentConstructor(compCtorFunc interface{}) fx.Option {
+	inFxType, outFxType, outEmbeds, err := constructFxInAndOut(compCtorFunc)
+	if err != nil {
+		return fx.Error(err)
+	}
+	// build reflect.Type of the constructor function that will be provided to `fx.Provide`
+	funcFxType := reflect.FuncOf([]reflect.Type{inFxType}, []reflect.Type{outFxType}, false)
+
+	hs := &compDef.HookStorage{}
+
+	// wrapper that receives fx-aware requirements, converts them into regular requirements, and calls the constructor
+	wrapperForConstructor := func(fxAwareReqs reflect.Value) reflect.Value {
+		constructorFunc := reflect.ValueOf(compCtorFunc)
+		// invoke the regular constructor with the correct arguments
+		ctorArgs := makeConstructorArgs(constructorFunc, fxAwareReqs, hs)
+		plainOuts := constructorFunc.Call(ctorArgs)
+		maybeAppendFxLifecycle(ctorArgs, fxAwareReqs)
+		// calling `constructFxInAndOut` earlier ensures that outs has exactly 1 element
+		return makeFxAwareProvides(plainOuts[0], outFxType, outEmbeds)
+	}
+
+	// constructor function value that will inform fx what the Components are
+	fxAwareProviderFunc := reflect.MakeFunc(funcFxType, func(args []reflect.Value) []reflect.Value {
+		reqs := args[0]
+		res := wrapperForConstructor(reqs)
+		return []reflect.Value{res}
+	})
+
+	fn := fxAwareProviderFunc.Interface()
+	return fx.Provide(fn)
+}
+
+func makeConstructorArgs(constructorFunc, fxAwareReqs reflect.Value, hs *compDef.HookStorage) []reflect.Value {
```

Comment:
Could we add doc for all functions in this file ? They're not exported but I think it will make future work on this easier.

Reply:
Added docs for all of the functions to explain what they do